### PR TITLE
feat: implement initial questions

### PR DESCRIPTION
**Pull request by friedrichwilken**
- change the endpoint to /conversations
- add a simple unit test for the endpoint, that will mock the controllers function calls

**Description**

Changes proposed in this pull request:

- implemented initialize conversation endpoint.
- Sample run: [link](https://github.com/kyma-project/kyma-companion/issues/29#issuecomment-2333638921)

**Related issue(s)**

- https://github.com/kyma-project/kyma-companion/issues/29

### CODE REVIEW

File: src/agents/initial_questions/inital_questions.py [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+
+    def __init__(self, model: IModel,
+                 prompt_template: str = INITIAL_QUESTIONS_PROMPT,
+                 output_parser: IOutputParser = QuestionOutputParser
+                 ) -> None:
+        self.model = model
+        self.prompt_template = prompt_template
+        self.output_parser = output_parser
+
+    def generate_questions(self, context: str) -> list[str]:
```
Comment (mfaizanse) [inline src/agents/initial_questions/inital_questions.py:42]: TODO: Unit test.

File: src/agents/initial_questions/inital_questions.py [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+        chain = LLMChain(
+            llm=self.model,
+            prompt=prompt,
+            output_parser=self.output_parser,
+        )
+
+        # Run the chain.
+        return chain.run(context=context)
+
+    def fetch_relevant_data_from_k8s_cluster(
```
Comment (mfaizanse) [inline src/agents/initial_questions/inital_questions.py:60]: TODO: Unit test.

File: src/agents/initial_questions/output_parser.py [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+from langchain_core.output_parsers import BaseOutputParser
+
+class IOutputParser(Protocol):
+    """Interface for OutputParser."""
+
+    def parse(self, output: str) -> list[str]:
+        """Parse the output and return the questions."""
+        ...
+
+class QuestionOutputParser(BaseOutputParser[list[str]]):
```
Comment (mfaizanse) [inline src/agents/initial_questions/output_parser.py:13]: TODO: Unit test.

File: src/agents/initial_questions/output_parser.py [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+    """OutputParser for InitialQuestionsAgent."""
+
+    def parse(self, output: str) -> list[str]:
```
Comment (mfaizanse) [inline src/agents/initial_questions/output_parser.py:16]: TODO: Unit test.

File: src/agents/memory/redis_checkpointer.py [hunk 5]
```diff
@@ -189,3 +208,35 @@ async def aput_writes(
         #  Currently, only necessary methods are
         #  implemented.
         return None
+
+    async def add_conversation_message(
```
Comment (mfaizanse) [inline src/agents/memory/redis_checkpointer.py:212]: TODO: Unit test.

File: src/agents/memory/redis_checkpointer.py [hunk 6]
```diff
@@ -189,3 +208,35 @@ async def aput_writes(
...
+
+        try:
+            async with get_async_connection(self.async_connection) as conn:
+                key = f"history:{conversation_id}"
+                await conn.lpush(key, message.model_dump_json())
+
+        except Exception as e:
+            raise Exception(f"Failed to add conversation message: {str(e)}") from e
+
+    async def get_all_conversation_messages(
```
Comment (mfaizanse) [inline src/agents/memory/redis_checkpointer.py:228]: TODO: Unit test.

File: src/routers/conversations.py [hunk 7]
```diff
@@ -19,15 +28,65 @@ def get_conversation_service() -> IService:
...
     tags=["conversations"],
 )
 
 
-@router.get("/init")
-async def init() -> dict:
-    """Endpoint to initialize the chat with the Kyma companion"""
-    return {"message": "Chat is initialized!"}
+@router.post("/", response_model=InitialQuestionsResponse)
+async def init_conversation(
```
Comment (mfaizanse) [inline src/routers/conversations.py:37]: TODO: Unit test.

File: src/services/conversation.py [hunk 8]
```diff
@@ -16,53 +24,80 @@
 class IService(Protocol):
     """Service interface"""
 
-    def init_conversation(self) -> dict:
-        """Initialize the conversation"""
+    async def new_conversation(
```
Comment (mfaizanse) [inline src/services/conversation.py:27]: TODO: Unit test.

File: src/services/conversation.py [hunk 9]
```diff
@@ -16,53 +24,80 @@
...
-        self.kyma_graph = KymaGraph(model, memory)
-
-    def init_conversation(self) -> dict:
+        self.model_factory = ModelFactory()
+        self.model = self.model_factory.create_model(LLM.GPT4O_MODEL)
+        self.memory = RedisSaver(async_connection=initialize_async_pool(url=REDIS_URL))
+        self.kyma_graph = KymaGraph(self.model, self.memory)
+        self.init_questions_agent = InitialQuestionsAgent(model=self.model)
+
+    async def new_conversation(
```
Comment (mfaizanse) [inline src/services/conversation.py:59]: TODO: Unit test.

File: src/services/conversation.py [hunk 10]
```diff
@@ -16,53 +24,80 @@
...
+                type=QueryType.INITIAL_QUESTIONS,
+                query="",
+                response="\n".join(questions),
+                timestamp=time.time(),
+            ),
+        )
+
+        return questions
+
+    async def generate_initial_questions(
```
Comment (mfaizanse) [inline src/services/conversation.py:83]: TODO: Unit test.

File: src/services/k8s.py [hunk 11]
```diff
@@ -0,0 +1,257 @@
...
+                result.append(event)
+
+        return result
+
+
+class DataSanitizer:
+    """Sanitize the data by from Kubernetes resources removing sensitive information."""
+
+    @staticmethod
+    def sanitize(data: dict | list[dict]) -> dict | list[dict]:
```
Comment (mfaizanse) [inline src/services/k8s.py:236]: TODO: Unit test.

File: src/utils/utils.py [hunk 12]
```diff
@@ -11,3 +12,12 @@ def create_ndjson_str(obj: dict) -> str:
         str: A stringified Newline-delimited JSON.
     """
     return f"{json.dumps(obj)}\n"
+
+
+def create_session_id() -> str:
```
Comment (mfaizanse) [inline src/utils/utils.py:17]: TODO: Unit test.

File: src/agents/initial_questions/inital_questions.py (hunk 2 above)
Comment (mfaizanse) [reply src/agents/initial_questions/inital_questions.py:60]: Done

File: src/agents/initial_questions/output_parser.py (hunk 4 above)
Comment (mfaizanse) [reply src/agents/initial_questions/output_parser.py:16]: done

File: src/routers/conversations.py (hunk 7 above)
Comment (friedrichwilken) [reply src/routers/conversations.py:37]: done

File: src/services/conversation.py (hunk 9 above)
Comment (friedrichwilken) [reply src/services/conversation.py:59]: done

**Review by the1bit — APPROVED** [review]

**Pull request merged**